BraveWallet/Crypto/Stores/SwapTokenStore.swift

```diff
@@ -32,48 +33,384 @@ public class SwapTokenStore: ObservableObject {
+    switch base {
+    case .perSellAsset:
+      var decimal = 18
+      if let fromToken = selectedFromToken {
+        decimal = Int(fromToken.decimals)
+      }
+      let decimalString = weiFormatter.decimalString(for: respnose.buyAmount, decimals: decimal) ?? ""
+      if let bv = BDouble(decimalString) {
+        buyAmount = bv.decimalDescription
+      }
+    case .perBuyAsset:
+      var decimal = 18
+      if let toToken = selectedToToken {
+        decimal = Int(toToken.decimals)
+      }
+      let decimalString = weiFormatter.decimalString(for: respnose.sellAmount, decimals: decimal) ?? ""
+      if let bv = BDouble(decimalString) {
+        sellAmount = bv.decimalDescription
+      }
+    }
+    
+    if let bv = BDouble(respnose.price) {
+      // will need to invert price if price quote is based on buyAmount
+      // ref from slack:
+      // https://bravesoftware.slack.com/archives/C023VS4HJ6Q/p1636579425364500?thread_ts=1636570735.354500&cid=C023VS4HJ6Q
+      let price = base == .perSellAsset ? bv : 1 / bv
+      selectedFromTokenPrice = price.decimalDescription
+    }
+    
+    checkBalanceShowError(swapResponse: respnose)
+  }
+  
+  private func activeErc20Allowance(_ spenderAddress: String) {
+    let weiFormatter = WeiFormatter(decimalFormatStyle: .decimals(precision: 18))
+    guard
+      let fromToken = selectedFromToken,
+      let accountInfo = accountInfo,
+      let balanceInWeiHex = weiFormatter.weiString(
+        from: selectedFromTokenBalance?.decimalDescription ?? "",
+        radix: .hex,
+        decimals: Int(fromToken.decimals)
+      )
+    else { return }
+    transactionController.makeErc20ApproveData(
+      spenderAddress,
+      amount: balanceInWeiHex
+    ) { [weak self] success, data in
+        guard success else { return }
+        let txData = BraveWallet.TxData(
+          nonce: "",
+          gasPrice: "",
+          gasLimit: "",
+          to: fromToken.contractAddress,
+          value: "0x0",
+          data: data
+        )
+      self?.transactionController.addUnapprovedTransaction(
+          txData,
+          from: accountInfo.address,
+          completion: { success, txMetaId, error in
+            // should be observed
+          }
+      )
+    }
+  }
+  
+  private func gweiStringToDecimal(_ value: String) -> BDouble? {
+    guard !value.isEmpty, let bv = BDouble(value) else {
+      return nil
+    }
+    return bv / (BDouble(10) ** 9)
+  }
+  
+  private func checkBalanceShowError(swapResponse: BraveWallet.SwapResponse) {
+    guard
+      let accountInfo = accountInfo,
+      let sellAmountValue = BDouble(sellAmount),
+      let gasLimit = gweiStringToDecimal(swapResponse.estimatedGas),
+      let gasPrice = gweiStringToDecimal(swapResponse.gasPrice),
+      let fromToken = selectedFromToken
+    else { return }
+    
+    // Check if balance is insufficient
+    if sellAmountValue > (selectedFromTokenBalance ?? 0) {
```

Comment:
Should we error out if a balance doesn't exist yet? (0 fallback will basically mean any sell amount value will trigger this error)